requests-for-comments/2022-11-08-supported-languages.md

```diff
@@ -0,0 +1,169 @@
+# Request for comments: Supported languages at PostHog
+
+Since day 1 here at Posthog we've supported two and half core languages: Python and Javascript/Typescript.
+
+This set of tools has carried us a long way and we should both the cpython runtime and node a huge hats off for bringing us this far.
+
+## Problem statement
+
+We have the opportunity coming up to rebuild or greenfield build out services that are critical parts of our data pipeline. It will be important for these services to be correct, fast, and efficient. Considering this now is a good time to ask: Are we using the correct tools for the job?
```

Comment:
I'd suggest some of the benefits of switching language comes from re-writing/extracting the service and only part from re-writing/extracting the service _in a different language

At a previous job we spent a while rewriting from Python to C# because "C# is faster than Python". Then realised that we were CPU bound and moved a bunch of work onto the GPU. Python and C# were then faster than we needed them to be.

So: "are we using the correct tools for the job" is a great question. But being really clear on why we're asking that and how we'll know when we're done is super important. E.g. moving ingestion onto the GPU is probably more complicated and not actually faster.

So, there are two interesting questions here:

1) what needs to be done to make ingestion safer/faster/misc
2) how many programming languages can we support

These are really different questions. 

I'm not working on ingestion so I can have opinions on 1 but "so what?"

On "how many programming languages"... The problem with a new (to you) programming language is almost never the language.

Aside from https://neverworkintheory.org/2014/01/29/stefik-siebert-syntax.html Python and Ruby are consistently measured as easiest to learn. C-style languages are no easier to learn than a language made of random keywords.

The problem with a new (to you) language is generally the tooling. Most noticeably, in my experience: dependency management, and building and releasing things.

(go update our android library from java 8 to java 19 if you want proof of this :))

So, we need a comparison of building, releasing, and running in k8s for the languages we consider. I think we can consider a smaller list (although C# marketing should be "Java but good", and I've a friend we could hire if we started using clojure which seems to inspire massive love from folk that use it)

The other thing is adoption within the org. Who needs to learn the language, and how and when do they do that? How do we know when to write it? Are we migrating to it or adding it alongside? Which services mustn't it be used for?

And finally: hiring. "Come work here because you're excited about language X. Incidentally the first few months you'll be working on these bugs in Python". So, we can pull from a wider pool with a wider pool of languages but are we in a place to hire someone who _only_ wants to work on language X

----

Bonus "after finally" point... who *owns* tooling for each language? Does the platform team commit to providing support for building, deploying, and running all the languages? Is that in our definition of platform? Or do we need some champions